src/atcalls.c

```diff
@@ -94,23 +95,23 @@ _fullpathat(int dirfd, const char *relative, char *buf)
 
 int faccessat(int dirfd, const char *pathname, int mode, int flags)
 {
-    struct attrlist al = {
-        .bitmapcount = ATTR_BIT_MAP_COUNT,
-        .commonattr  = ATTR_CMN_USERACCESS,
-    };
-    struct {
-        uint32_t length;
-        uint32_t access;
-    } __attribute__((aligned(4), packed)) ab;
-    unsigned long opts = 0;
-
-    ERR_ON(EINVAL, flags & ~AT_SYMLINK_NOFOLLOW);
-    if (flags & AT_SYMLINK_NOFOLLOW)
-        opts |= FSOPT_NOFOLLOW;
-    if (getattrlistat(dirfd, pathname, &al, &ab, sizeof(ab), opts) < 0)
-        return -1;
-    ERR_ON(EACCES, (ab.access & mode) != (uint32_t)mode);
-    return 0;
+    ERR_ON(EINVAL, flags & ~AT_EACCESS);
+    uid_t ruid = getuid(), euid = geteuid();
+    gid_t rgid = getgid(), egid = getegid();
+    int check_euid = ruid != euid && (flags & AT_EACCESS);
+    int check_egid = rgid != egid && (flags & AT_EACCESS);
+    if (check_euid)
+        setreuid(euid, ruid);
+    if (check_egid)
+        setregid(egid, rgid);
+    int access_rc = ATCALL(dirfd, pathname, access(pathname, mode));
+    int access_errno = errno;
+    if (check_euid)
+        setreuid(ruid, euid);
+    if (check_egid)
+        setregid(rgid, egid);
+        errno = access_errno;
+    return access_rc;
```

Comment:
Interesting idea to emulate that kernel does.

Any way a few remarks.
1. You should respect return code of `setreuid` / `setregid`;
2. You should close a descriptor which is returned by `dirfd`;
3. You may simplify condition like `if (check_euid)` to `if (ruid != euid)` because `flags` should always have `AT_EACCESS`.

Reply:
1. I disagree. According to the manual entries for `setreuid` and `setregid`, they won't fail when swapping real and effective ids. It'll be much uglier and serve no purpose. But I'll change it if that doesn't convince you.
2. `dirfd` in this instance isn't the `dirfd` function that returns a file descriptor from a `DIR` object. It's the name of the first parameter to the `faccessat` function and it should not be closed.
3. That's not true. `flags` can equal 0.

Reply:
Actually, (1) wouldn't be too ugly. We could return `-1` if the first `setreuid` failed, and `abort` if any of the other three calls failed. It's the logic to undo previous successful calls before returning `-1` that would make it ugly (and handling subsequent hypothetical failures while attempting to do that). I don't like the idea of having code in there to abort, but it's technically OK because it can never be executed, and it's the only appropriate response to such a failure. We can't have any possibility of returning with uids/gids different to what they were upon entry to the function (since we're talking about the possibility of these calls failing). But having calls to `abort` in there could be misleading to anyone reading the code in the future.

It would look like this:

    int faccessat(int dirfd, const char *pathname, int mode, int flags)
    {
        ERR_ON(EINVAL, flags & ~AT_EACCESS);
        uid_t ruid = getuid(), euid = geteuid();
        gid_t rgid = getgid(), egid = getegid();
        int check_euid = ruid != euid && (flags & AT_EACCESS);
        int check_egid = rgid != egid && (flags & AT_EACCESS);
        if (check_euid)
            if (setreuid(euid, ruid) < 0)
                return -1; // This can't happen
        if (check_egid)
            if (setregid(egid, rgid) < 0)
                abort(); // This can't happen
        int access_rc = ATCALL(dirfd, pathname, access(pathname, mode));
        int access_errno = errno;
        if (check_euid)
            if (setreuid(ruid, euid) < 0)
                abort(); // This can't happen
        if (check_egid)
            if (setregid(rgid, egid) < 0)
                abort(); // This can't happen
            errno = access_errno;
        return access_rc;
    }

I don't like this, but I'll do it if you insist.

Reply:
Sorry, I do have a lot of background stress and issues and I was complely worng about `dirfd`; yes, you're right it isn't syscall name. I was confused :)

I love your last approach, but may I ask you to fix formating for `errno = access_errno;` part to avoid one level of tabulation which misslead at reading.

Reply:
Thanks. That's fixed.